mabwiser/treebandit.py

```diff
@@ -0,0 +1,163 @@
+# -*- coding: utf-8 -*-
+# SPDX-License-Identifier: Apache-2.0
+
+from copy import deepcopy
+from typing import Union, Dict, List, NoReturn, Optional, Callable
+
+import numpy as np
+from sklearn.tree import DecisionTreeClassifier
+
+from mabwiser.base_mab import BaseMAB
+from mabwiser.greedy import _EpsilonGreedy
+from mabwiser.linear import _Linear
+from mabwiser.popularity import _Popularity
+from mabwiser.rand import _Random
+from mabwiser.softmax import _Softmax
+from mabwiser.thompson import _ThompsonSampling
+from mabwiser.ucb import _UCB1
+from mabwiser.utils import reset, argmax, Arm, Num, _BaseRNG
+
+
+class _TreeBandit(BaseMAB):
+    def __init__(self, rng: _BaseRNG, arms: List[Arm], n_jobs: int, backend: Optional[str],
+                 lp: Union[_EpsilonGreedy, _Linear, _Popularity, _Random, _Softmax, _ThompsonSampling, _UCB1],
+                 tree_parameters: Dict):
+        super().__init__(rng, arms, n_jobs, backend)
+        self.lp = lp
+        self.tree_parameters = tree_parameters
+        self.tree_parameters["random_state"] = rng.seed
+
+        self.arm_to_tree = None
+        self.arm_to_rewards = None
+
+        # Initialize the arm expectations to nan
+        # When there are neighbors, expectations of the underlying learning policy is used
+        # When there are no neighbors, return nan expectations
+        reset(self.arm_to_expectation, np.nan)
+
+    def fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+
+        # Reset the decision trees of each arm
+        self.arm_to_tree = {arm: DecisionTreeClassifier(**self.tree_parameters) for arm in self.arms}
+        self.arm_to_rewards = {arm: dict() for arm in self.arms}
+
+        if isinstance(self.lp, _ThompsonSampling) and self.lp.binarizer:
+            self.lp.is_contextual_binarized = False
+            rewards = self.lp._get_binary_rewards(decisions, rewards)
+            self.lp.is_contextual_binarized = True
+
+        # Calculate fit
+        self._parallel_fit(decisions, rewards, contexts)
+
+    def partial_fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+
+        # Update rewards list at leaves
+        for i, arm in enumerate(decisions):
+            leaf_index = self.arm_to_tree[arm].apply([contexts[i]])[0]
+            self.arm_to_rewards[arm][leaf_index] = np.append(self.arm_to_rewards[arm][leaf_index], rewards[i])
```

Comment:
what is the data structure for arm_to_rewards[arm][index] .. is it a List or numpy Array? 

Cannot we do self.arm_to_rewards[arm][leaf_index].append(rewards[i])?

Reply:
I meant for them to be np arrays because later when we're using indexing like arm_rewards[leaf_indices == index], I think it only works with np arrays. And I don't think just .append() to works with np arrays.